src/fileio/datafile.cxx

```diff
@@ -928,6 +928,33 @@ void Datafile::add(Vector3D &f, const char *name, bool save_repeat) {
   }
 }
 
+namespace {
+// Read a value from file and check it matches reference_value, throw if not
+void checkGridValue(DataFormat* file, const std::string& name,
+                    const std::string& filename, const int reference_value) {
+  int file_value;
+  if (!file->read(&file_value, name)) {
+    throw BoutException("Could not read {} from file '{}'", name, filename);
+  }
+
+  if (file_value != reference_value) {
+    throw BoutException("{} ({}) in file '{}' does not match value in mesh ({})", name,
+                        file_value, filename, reference_value);
+  }
+}
+
+// Check that the array sizes in \p file match those in existing \p mesh
+void checkFileGrid(DataFormat* file, const std::string& filename, const Mesh* mesh) {
+  checkGridValue(file, "MXG", filename, mesh->xstart);
+  checkGridValue(file, "MYG", filename, mesh->ystart);
+  checkGridValue(file, "MZG", filename, mesh->zstart);
+  // nx includes boundaries
+  checkGridValue(file, "nx", filename, mesh->GlobalNx);
+  checkGridValue(file, "ny", filename, mesh->GlobalNyNoBoundaries);
+  checkGridValue(file, "nz", filename, mesh->GlobalNzNoBoundaries);
```

Comment:
These two methods were added in v5 -- was this just me being lazy and not backporting them, or did we have a reason to not do so?

Reply:
We made the naming of all the global indices consistent in v5. I think it doesn't make sense to back-port because in v4 GlobalNy is without boundaries, and we don't want to add yet another member for 'GlobalNyWithBoundaries'. We could add the Global*NoBoundaries and just live with inconsistent definitions of GlobalNx/GlobalNy, but I think that's more likely to add to confusion... We can use:
```suggestion
  checkGridValue(file, "ny", filename, mesh->GlobalNy);
  checkGridValue(file, "nz", filename, mesh->LocalNz);
```